Dockerfile

```diff
@@ -102,12 +98,30 @@ WORKDIR /src
 ARG TESTPKGS
 RUN --mount=type=cache,target=/root/.cache/go-build --mount=type=cache,target=/go/pkg --mount=type=cache,target=/tmp go test -v -covermode=atomic -coverprofile=coverage.txt -coverpkg=${TESTPKGS} -count 1 ${TESTPKGS}
 
-FROM scratch AS kres-linux-amd64
-COPY --from=kres-linux-amd64-build /kres-linux-amd64 /kres-linux-amd64
+FROM embed-generate AS embed-abbrev-generate
+WORKDIR /src
+RUN echo -n 'undefined' > internal/version/data/sha && \
+    echo -n 'undefined' > internal/version/data/tag
 
 FROM scratch AS unit-tests
 COPY --from=unit-tests-run /src/coverage.txt /coverage-unit-tests.txt
 
+# cleaned up specs and compiled versions
+FROM scratch AS generate
+COPY --from=embed-abbrev-generate /src/internal/version internal/version
+
+# builds kres-linux-amd64
+FROM base AS kres-linux-amd64-build
+COPY --from=generate / /
+COPY --from=embed-generate / /
```

Comment:
does this mean we get `undefined`, the build should copy the generate only I believe

Reply:
The idea is copied from Talos actually - that is, if we always generate and write latest SHA and Tag we always get a dirty working directory. So instead of creating infinite dependency cycle we use placeholders during `generate` and write actual values of those during `build` so they never escape to the working directory.

Reply:
correct, but build should use the correct TAG and SHA and not from embed-generate, Talos does the same

Reply:
Talos goes like this during `generate` route: `embed-generate` -> `embed-abbrev-generate` (overwrite with `undefined`) -> `embed-abbrev` -> `generate`. So the actual (correct) values a overwritten with undefined and never escape the docker build process. During `build` it goes `embed-generate` -> `embed` -> `build-go` so we get the correct correct TAG and SHA and never overwrite them. Here logic is (mostly) the same.

You can actually verify it by making kres first and then calling `kres version`. If it was build using `make kres` it will always print correct SHA and TAG, and if it was build using `go build` and such, it will print undefined for both TAG and SHA.

Reply:
ahh makes sense, it was confusing reading only the PR changes :+1:

Reply:
Added more explanation into the PR description. Thanks!